### ISSUE TITLE
[charts/linkerd-control-plane] Allow .Values.identity.issuer.tls.* to be passed as existingSecret

### ISSUE THREAD

**Issue by brsolomon-deloitte** [linkerd/linkerd2#10229]
### What problem are you trying to solve?

Currently, installing `linkerd-control-plane` Helm chart requires passing literal string or file values for `.Values.identity.issuer.tls.crtPEM` and `.Values.identity.issuer.tls.keyPEM`. This is not ideal for GitOps-based systems where all reference state should be stored in Git.

### How should the problem be solved?

As the admin conducting the Helm install, it would be preferable to be able to pass `linkerd-control-plane` some type of `existingSecret` parameter that points to a `Secret` such as `linkerd-identity-issuer` (which is currently templated-out in `templates/identity.yaml`) that already exists in the target cluster.

```
identity:
  issuer:
    tls:
      existingSecret: my-existing-secret-name
```

If the parameter is passed, then `templates/identity.yaml` should simply skip templating-out the `Secret` called `linkerd-identity-issuer` that uses the literal values of `.Values.identity.issuer.tls.crtPEM` and `.Values.identity.issuer.tls.keyPEM`.

https://github.com/linkerd/linkerd2/blob/032845c846fe2e5d12922f0cf3b291b70013f802/charts/linkerd-control-plane/templates/identity.yaml#L8

This `Secret` could either be mandated to contain `crt.pem` and `key.pem`, or a "key mapping" could be made available too. This plays wells with GitOps using a mechanism such as `external-secrets-operator`.

The `existingSecret` pattern is commonly used, such as in the [Bitnami helm charts](https://github.com/bitnami/charts).

### Any alternatives you've considered?

https://github.com/linkerd/linkerd2/pull/6455 is related, but does not address this issue.

### Would you like to work on this feature?

maybe

**Comment by brsolomon-deloitte** [comment]
Also of note: it looks like the [`linkerd-viz` chart already does this](https://github.com/linkerd/linkerd2/blob/032845c846fe2e5d12922f0cf3b291b70013f802/viz/charts/linkerd-viz/values.yaml#L154), so it is an existing practice within at least some of the linkerd Helm charts too.

**Comment by whiskeysierra** [comment]
Am I correct to assume that it's this resource that needs to be made optional?

https://github.com/linkerd/linkerd2/blob/a37f632394cc85f757bde9493140b20d7ec47db2/charts/linkerd-control-plane/templates/identity.yaml#L3-L22

**Comment by brsolomon-deloitte** [comment]
> Am I correct to assume that it's this resource that needs to be made optional?

Yes, and then other resources that reference the secret by name (``) would need to make sure to conditionally use `{{ .Values.<foo>.existingSecret }}`.

**Comment by whiskeysierra** [comment]
Why? Wouldn`t `existingSecret` be a boolean and all the other usages would just refer to the secret by its pre-defined name, regardless of who created it?

**Comment by brsolomon-deloitte** [comment]
> Why? Wouldn't `existingSecret` be a boolean and all the other usages would just refer to the secret by its pre-defined name, regardless of who created it?

In my experience working with other Helm charts, it is more typical for `existingSecret` to be a string pointing to the name of a pre-existing `Secret` containing the existing credentials.

**Comment by whiskeysierra** [comment]
Ok, that would also work I guess. It's just different to the existing usages in the linkerd helm charts.

**Comment by whiskeysierra** [comment]
I wouldn't be surprised if the secrets that are being watched are hard coded.

**Comment by mateiidavid** [comment]
Sorry folks, took my time with a reply here. You can already do this in linkerd control plane, you just can't pick the name that you want. This shouldn't get in the way of any IaC pipelines. We have this documented, but I do think we could surface this a little bit better:

> It is important to note that the mechanism that Linkerd provides for setting issuer certificates and keys is also usable outside of cert-manager. Linkerd will read the linkerd-identity-issuer Secret, and if it’s of type kubernetes.io/tls, will use the contents as its TLS credentials. This means that any solution that is able to rotate TLS certificates by writing them to this secret can be used to provide dynamic TLS certificate management.

https://linkerd.io/2.12/tasks/automatically-rotating-control-plane-tls-credentials/#a-note-on-third-party-cert-management-solutions

In other words, if the secret is `kubernetes.io/tls` Linkerd will simply read it instead of creating it. A quick check from the CLI:

```
:; linkerd install --ignore-cluster | rg 'identity-issuer' -B 5
###
---
kind: Secret
apiVersion: v1
metadata:
  name: linkerd-identity-issuer
--
          allowPrivilegeEscalation: false
          seccompProfile:
            type: RuntimeDefault
        volumeMounts:
        - mountPath: /var/run/linkerd/identity/issuer
          name: identity-issuer
--
        volumeMounts:
        - mountPath: /run
          name: linkerd-proxy-init-xtables-lock
      serviceAccountName: linkerd-identity
      volumes:
      - name: identity-issuer
        secret:
          secretName: linkerd-identity-issuer
```

Notice we get the secret printed at the top when we grep. And now, if we change the scheme:

```
:; linkerd install --set 'identity.issuer.scheme=kubernetes.io/tls' --ignore-cluster | rg 'identity-issuer' -B 5
          allowPrivilegeEscalation: false
          seccompProfile:
            type: RuntimeDefault
        volumeMounts:
        - mountPath: /var/run/linkerd/identity/issuer
          name: identity-issuer
--
        volumeMounts:
        - mountPath: /run
          name: linkerd-proxy-init-xtables-lock
      serviceAccountName: linkerd-identity
      volumes:
      - name: identity-issuer
        secret:
          secretName: linkerd-identity-issuer
```

Most CAs give you the ability to specify the secret name when they need to create a keypair. I don't think the additional logic needed to customise the secret name is needed here, tbh. Changing the scheme should do what you want it to. We can make this solution more obvious by updating the helm docs to point it this out (in an explicit way). Thoughts?

**Comment by brsolomon-deloitte** [comment]
>  In other words, if the secret is kubernetes.io/tls Linkerd will simply read it instead of creating it.

No, this is simply not true as stated. Per the template, the `Secret/linkerd-identity-issuer` is templated out _under the following condition_:

https://github.com/linkerd/linkerd2/blob/a37f632394cc85f757bde9493140b20d7ec47db2/charts/linkerd-control-plane/templates/identity.yaml#L6

What is the implicit suggestion here - that a user sets something like:

```yaml
identity:
  issuer:
    # force template check to evaluate to false so Secret does not get auto-created
    scheme: does-not-match
```

just to force the conditional to evaluate the `false`??

They can't set `identity.issuer=null` since they would logically want to retain values such as `issuanceLifetime` at their defaults.

**Comment by mateiidavid** [comment]
Closed through #10349. Helm docs should be a bit more explicit as to what values you have to set to either provide an existing CA, or an existing issuer. Thanks for the conversation folks :)
